unic/char/property/src/macros.rs

```diff
@@ -108,62 +147,220 @@ macro_rules! char_property {
-                        stringify!($long) => Ok($name::$variant),
-                    )*
-                    $(
-                        str if ::std::ascii::AsciiExt::eq_ignore_ascii_case(str, stringify!($abbr))
-                            => Ok($name::$variant),
-                        str if ::std::ascii::AsciiExt::eq_ignore_ascii_case(str, stringify!($long))
-                            => Ok($name::$variant),
-                    )*
-                    _ => Err(()),
-                }
-            }
+        char_property! {
+            __impl FromStr for $name; $(
+                $abbr => $name::$variant;
+                $long => $name::$variant;
+            )*
         }
 
-        impl $crate::CharProperty for $name {
-            fn prop_abbr_name() -> &'static str { $abbr_name }
-            fn prop_long_name() -> &'static str { $long_name }
-            fn prop_human_name() -> &'static str { $human_name }
+        char_property! {
+            __impl CharProperty for $name;
+            $abbr_name;
+            $long_name;
+            $human_name;
+        }
+
+        char_property! {
+            __impl Display for $name by EnumeratedCharProperty
         }
 
         impl $crate::EnumeratedCharProperty for $name {
             fn all_values() -> &'static [$name] {
-                const VALUES: &[$name] = &[
-                    $($name::$variant,)+
-                ];
+                const VALUES: &[$name] = &[ $($name::$variant,)+ ];
                 VALUES
             }
-
             fn abbr_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($abbr), )*
-                }
+                match *self { $( $name::$variant => stringify!($abbr), )* }
             }
-
             fn long_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($long), )*
-                }
+                match *self { $( $name::$variant => stringify!($long), )* }
             }
-
             fn human_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => $human, )*
+                match *self { $( $name::$variant => $human, )* }
+            }
+        }
+    };
+
+    // == Binary Property == //
+
+    (
+        $(#[$name_meta:meta])*
+        pub struct $name:ident(bool) {
+            abbr => $abbr_name:expr;
+            long => $long_name:expr;
+            human => $human_name:expr;
+
+            data_table_path => $data_path:expr;
+
+            pub fn $bool_name:ident(&self) -> bool;
+        }
+
+        pub fn $shorthand_name:ident(char) -> bool;
+
+        $(#[$char_trait_meta:meta])*
+        pub trait $char_trait_name:ident for char {
+            $(#[$char_trait_fn_meta:meta])*
+            pub fn $char_trait_fn_name:ident(self) -> bool;
+        }
+
+        $(#[$str_trait_meta:meta])*
+        pub trait $str_trait_name:ident for str {
```

Comment:
Does it make sense for _all_ binary properties to have a trait for `str`? And if we do, wouldn't it make sense just to reuse one trait for both? (So a (strawman) `UnicodeLowercase` trait that offers the fn `is_unicode_lowercase`, which is impl'd for both `char` and `str`.)

Some cases will definitely want a `str` impl. But I think some wouldn't: does `string.is_question_marks` seem like useful functionality, or something that's better left out of the API and just expressed as the marginally more complicated `string.chars().all(|c| c.is_question_mark())`?